src/kbmod/image_collection.py

```diff
@@ -0,0 +1,684 @@
+        if "std_idx" not in metadata.columns:
+            no_std_map = True
+            self.data["std_idx"] = [None]*len(self.data)
+
+        if "ext_idx" not in metadata.columns:
+            no_std_map = True
+            self.data["ext_idx"] = [None]*len(self.data)
+
+        # standardizers are not Falsy - a list of empty lists, Nones, empty
+        # tuples, empty arrays etc...
+        if (standardizers is not None and any(standardizers)) and no_std_map:
+            std_idxs, ext_idxs = [], []
+            for i, stdFits in enumerate(standardizers):
+                for j, ext in enumerate(stdFits.exts):
+                    std_idxs.append(i)
+                    ext_idxs.append(j)
+            self.data["std_idx"] = std_idxs
+            self.data["ext_idx"] = ext_idxs
+
+        self._userColumns = [
+            col for col in self.data.columns
+            if col not in self._supporting_metadata
+        ]
+
+    @classmethod
+    def read(cls, *args, format=None, units=None, descriptions=None, **kwargs):
+        """Create ImageCollection from a file containing serialized image
+        collection.
+
+        Parameters
+        ----------
+        filepath : `str`
+            Path to the file containing the serialized image collection.
+
+        Returns
+        -------
+        ic : `ImageCollection`
+            Image Collection
+        """
+        metadata = Table.read(*args, format=format, units=units,
+                              descriptions=descriptions, **kwargs)
+        metadata["wcs"] = [WCS(w) for w in metadata["wcs"] if w is not None]
+        metadata["bbox"] = [json.loads(b) for b in metadata["bbox"]]
+        metadata["config"] = [json.loads(c) for c in metadata["config"]]
+        meta = json.loads(metadata.meta["comments"][0],)
+        metadata.meta = meta
+        return cls(metadata)
+
+    @classmethod
+    def fromStandardizers(cls, standardizers, meta=None):
+        """Create ImageCollection from a collection `Standardizers`.
+
+        The `Standardizer` is "unravelled", i.e. the shared metadata is
+        duplicated for each entry marked as processable by the standardizer. On
+        an practical example - MJD timestamps are shared by all 62 science
+        exposures created by DECam in a single exposure, but pointing of each
+        one isn't. So each pointing is a new row in the metadata table for each
+        of the individual pointings.
+
+        Parameters
+        ----------
+        standardizers : `iterable`
+            Collection of `Standardizer` objects.
+
+        Returns
+        -------
+        ic : `ImageCollection`
+            Image Collection
+        """
+        unravelledStdMetadata = []
+        for i, stdFits in enumerate(standardizers):
+            # needs a "validate standardized" method here or in standardizers
+            stdMeta = stdFits.standardizeMetadata()
+
+            # unravel all standardized keys whose values are iterables unless
```

Comment:
Writing a different standardizer is like having a user define this for us. This proposal is assuming that all the data required can be located in the PRIMARY header and the selected HDU extension. This is not the case in general.

In a simplistic description, Rubin defines their data products as collection of a single CCD exposure alongside with their definitions of `SkyWcs`, `PSF`, mask, variance, catalog metadata etc. The DECam CommunityPipelines define their data products as collection of images from all CCDs at the same time. For CommunityPipe data products the variance, mask, PSF etc. are all different data products that live separately from their "image" data products (raw, calibrated, coadded imdiffed etc.). For example such as in a different FITS file or in an online service-like database thing. 

For different Sci Pipe and Community Pipe data products these "CCD"s include usable and not-usable pixels (such as overscan and bias estimation regions for example) and sometimes not. This sometimes makes it so that one can calculate the variance and bias planes from the data they already have, and sometimes they might just have to go and fetch the additional data they need. Sometimes they may have already fetched that data and have it on disk somewhere. A lot of the times, this is actually (annoyingly) encoded in names of files - for example SDSS relies on that to figure out what's what. This means that sometimes to figure out what file/access mechanism we need, we would need to look at more than just the primary header and the given HDU or provide a definition that's a convention in nature.

So to sum it up the data products we get we have to observe holistically because standardizers encode rules and conventions that differ from a data source to data source, and more specifically user to user. Standardizers encode not just the conventions, but the mechanism by which those conventions are translated to internal KBMOD abstractions (f.e. a numpy arrays of zeros and ones for masks).